common/lib/transport/connectionmanager.js

```diff
@@ -652,13 +660,16 @@ var ConnectionManager = (function() {
 		 * on a new connection, with implications for msgSerial and channel state */
 		var self = this;
 		connectionSerial = (connectionSerial === undefined) ? -1 : connectionSerial;
-		/* Note that this is also run on clean connections; the msgSerial is a
-		 * noop, but the channel reattach is needed for channels that were
-		 * previously in the attached state even though the connection mode was
-		 * 'clean' due to a freshness check - see https://github.com/ably/ably-js/issues/394 */
-		if(this.connectionId !== connectionId)  {
-			Logger.logAction(Logger.LOG_MINOR, 'ConnectionManager.setConnection()', 'New connectionId; resetting msgSerial and reattaching any attached channels');
+		/* If no previous connectionId, don't reset the msgSerial as it may have been set by recover data */
+		if(this.connectionId && (this.connectionId !== connectionId))  {
+			Logger.logAction(Logger.LOG_MINOR, 'ConnectionManager.setConnection()', 'Resetting msgSerial');
 			this.msgSerial = 0;
+		}
+		/* but do need to reattach channels, for channels that were previously in
+		 * the attached state even though the connection mode was 'clean' due to a
+		 * freshness check - see https://github.com/ably/ably-js/issues/394 */
```

Comment:
@SimonWoolf is this behaviour / bug definitely covered by the spec?  I suspect other libraries could have this same bug and thus attached channels prior to losing connection state will remain connected in the lib.

Reply:
@mattheworiordan it isn't in the spec because it's an artifact of the freshness check, which isn't yet in the spec -- the docs issue to do so is https://github.com/ably/docs/issues/287